### PR TITLE
Prevent full page re-renders on URL hash changes

### LINKED ISSUE #9252: Internal hash links on article pages cause full page rerenders
**Describe the bug**
When using an internal hash link on article pages the full page rerender. This is noticeable when dynamically loaded content is on the page such as embeds.

hash-bug

**To Reproduce**
Steps to reproduce the behaviour:

1. Go to https://www.test.bbc.com/mundo/articles/c20myn5q195o
2. Tab to the header skip link and hit enter
3. Observe the twitter embeds reloading

**Expected behaviour**
Navigating to sections of the article should not reload the full article page.

**Testing notes**

- [x] This bug fix is expected to need manual testing.

**Checklist**
- [ ] (BBC contributors only) This issue follows the [repository use guidelines](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/repository-guidelines.md)

**Additional context**

### PR DESCRIPTION
**Pull request by amoore108**
Resolves #9252

**Overall change:**
Adds logic to prevent unnecessary re-renders of the pages when the user uses 'skip to content' links. Previously when a user used a 'skip to content' link, the page would re-render. This is generally unnoticeable, however on pages that use Social embeds, this is more obvious (as seen in the linked ticket)

The app should behave as normal in almost all circumstances, only in instances where a `#` is added to the end of the URL will the app attempt to prevent a re-render of the page.

**Code changes:**
- Bulk of work was performed by @jroebu14 in https://github.com/bbc/simorgh/pull/9274, this PR makes a few changes to the equality checks in the `isHashChange` file. Previously, equality checks were performed on the entire state object, which could result in performance issues. This PR removes those checks and checks only for `location.hash` changes. 
- The `withHashChangeHandler` HOC is used on every page, so testing of the more subtle state changes on pages should be performed to ensure the memoisation hasn't affected anything

---

- [x] I have assigned myself to this PR and the corresponding issues
- [x] I have added the `cross-team` label to this PR if it requires visibility across World Service teams
- [x] I have assigned this PR to the Simorgh project
- [x] (BBC contributors only) This PR follows the [repository use guidelines](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/repository-guidelines.md)

**Testing:**

- [ ] Automated (jest and/or cypress) tests added (for new features) or updated (for existing features)
- [ ] If necessary, I have run the local E2E non-smoke tests relevant to my changes (`CYPRESS_APP_ENV=local CYPRESS_SMOKE=false yarn test:e2e:interactive`)
- [ ] This PR requires manual testing

### CODE REVIEW

File: src/app/containers/PageHandlers/withHashChangeHandler/index.test.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
+import isHashChange from './isHashChange';
+
+describe('isHashChange', () => {
+  describe('prevent a rerender', () => {
+    it('should return true if the location.hash prop changes, preventing a re-render', () => {
```
Comment (ryanmccombe) [inline src/app/containers/PageHandlers/withHashChangeHandler/index.test.js]: Wondering if there is a very niche scenario where someone clicks skip to content, and then does a client-side navigation.  In that scenario, `location.hash` changes, but also `location.pathname`.  Might this prevent a page from rerendering when it actually needs to?

We do barely any client side navigation at the minute, but it is being used between article pages - eg being on a `:service/articles/:id` page and clicking on another `:service/articles/:id` link will be done client side.

Comment (ryanmccombe) [reply src/app/containers/PageHandlers/withHashChangeHandler/index.test.js]: yeah was able to reproduce this issue, eg:

1. update this file locally: https://github.com/bbc/simorgh/blob/latest/data/afaanoromoo/articles/c4g19kgl85ko.json#L171 to be: `"locator": "/amharic/articles/czqverekrldo",` 
2. visit http://localhost:7080/afaanoromoo/articles/c4g19kgl85ko#content
3. click the link in the first paragraph

Comment (amoore108) [reply src/app/containers/PageHandlers/withHashChangeHandler/index.test.js]: Ahh, good catch. I can reproduce that as well. I'll have a further look to see if we can mitigate that for the pathname as well. Wasn't sure what the client-side routing situation was, but good to know its used in some areas.

Comment (amoore108) [reply src/app/containers/PageHandlers/withHashChangeHandler/index.test.js]: I've added in a `pathname` check and reworded the function slightly to better reflect its functionality.

Appears to resolve the issue you found, although slightly concerned that there are more of these gotchas if the URLs are used for much more than routing.

**Review by ryanmccombe — APPROVED** [review]
👍

File: src/app/containers/PageHandlers/withHashChangeHandler/index.test.js (hunk 1 above)
Comment (ryanmccombe) [reply src/app/containers/PageHandlers/withHashChangeHandler/index.test.js]: LGTM 👍 only /articles links use CSR at the minute - rolling it out further has been on the backlog for a while but never made it to the top

**Review by nicola-d-bbc — APPROVED** [review]

File: src/app/containers/PageHandlers/withHashChangeHandler/isHashChangeOnSamePath.js [hunk 2]
```diff
@@ -0,0 +1,10 @@
+export default (
```
Comment (jroebu14) [inline src/app/containers/PageHandlers/withHashChangeHandler/isHashChangeOnSamePath.js:1]: do you think this covers the case if some other props change at the same time as a hash change? I _think_ that's what the previous implementation checked for.

Comment (jroebu14) [reply src/app/containers/PageHandlers/withHashChangeHandler/isHashChangeOnSamePath.js:1]: so you would want to check that the **only** prop that has changed is the hash prop

Comment (amoore108) [reply src/app/containers/PageHandlers/withHashChangeHandler/isHashChangeOnSamePath.js:1]: I've done a bit of testing on this by updating a state value at the top of the component tree and passing it down to the `ArticlePage` component as a prop and it appears to update as expected. The social embeds re-render (causing that unstyled appearance at least for Twitter) in this case, which I think is what we want to happen as the previously applied memoisation is not preventing re-renders from happening.

Comment (jroebu14) [reply src/app/containers/PageHandlers/withHashChangeHandler/isHashChangeOnSamePath.js:1]: cool thanks for testing this

**Review by jroebu14 — APPROVED** [review]

**Pull request merged**